hoomd/pytest/test_gsd.py

```diff
@@ -0,0 +1,378 @@
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
+
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 mode='xb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+    with pytest.raises(Exception):
+        sim.run(1)
+
+    # test mode=xb creating file
+    filename_xb = tmp_path / "temporary_test_file_xb.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params[('p1', 'p1')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
+
+    gsd_writer = hoomd.write.GSD(filename=filename_xb,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 mode='xb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    sim.run(5)
+
+    with gsd.hoomd.open(name=filename_xb, mode='rb') as traj:
+        for step in range(5):
+            assert_equivalent_snapshots(traj[step], snapshot_list[step])
+
+
+def test_write_gsd_null_filter(simulation_factory, device,
+                               hoomd_snapshot, tmp_path):
+
+    # test Null filter
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params[('p1', 'p1')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
+
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 filter=hoomd.filter.Null(),
+                                 mode='wb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    sim.run(3)
+
+    with gsd.hoomd.open(name=filename, mode='rb') as traj:
+        for step in range(3):
+            assert traj[step].particles.N == 0
+
+
+def test_write_gsd_type_filter(simulation_factory, device,
```

Comment:
I don't think this test is necessary, we already test filters elsewhere. but if you think we should keep this @joaander we can.

Reply:
In addition to whether to keep this test, I have a question about overwriting the same gsd file written with a different filter. Like I mentioned in the description of this PR, I was trying to test both two filters in one test session, but somehow the second ```gsd_writer``` with the type filter cannot overwrite the gsd file created by the ```gsd_writer``` with the null filter. I'm wondering if this is universally true with filters or it's just because I used the null filter first.

Reply:
Are you using the 'xb' mode? That will raise an error if the file exists. Use the 'wb' mode to overwrite files.

Reply:
I am using the 'wb' mode, and that's why I'm a little confused. I'm seeing a ```RuntimeError``` which says that the filename I provided is an invalid argument. The filename I provided is exactly the file created by the first ```gsd_writer```. When I changed this filename to a new one, it worked.

Reply:
I think it is fine to remove the type filter test. You have the null filter test and all the other tests implicitly check the all filter. That's enough to demonstrate that the `filter` argument has an effect.